airflow/jobs/scheduler_job.py

```diff
@@ -1417,6 +1420,42 @@ def _clean_tis_without_dagrun(self, session):
                     raise
             guard.commit()
 
+    @provide_session
+    def _missing_dag_file_cleanup(self, session: Session = None):
+        """Fails task instances and DagRuns of DAGs that no longer exist in the dag folder/DB"""
+        states_to_check = State.unfinished - frozenset([State.NONE, State.SHUTDOWN])
+        tis = session.query(TI).filter(TI.state.in_(states_to_check)).all()
+        missing_dags = {}
+        dag_runs = set()
+
+        for ti in tis:
+            if ti.dag_id in missing_dags:
+                ti.set_state(State.FAILED, session=session)
+                continue
+            # Dag no longer in dagbag?
+            if not self.dagbag.has_dag(ti.dag_id, session=session):
+                ti.set_state(State.FAILED, session=session)
+                dag_runs.add(ti.dag_run)
+                missing_dags[ti.dag_id] = [ti]
+                continue
+            # Dag file no longer exists?
+            if not os.path.exists(ti.dag_model.fileloc):
+                ti.set_state(State.FAILED, session=session)
+                dag_runs.add(ti.dag_run)
+                missing_dags[ti.dag_id] = [ti]
```

Comment:
I don't think this check is necessary given if the file doesn't exist it will (eventually, after a little while) no longer be in the dag.

@kaxil do we ever remove dags from the DagBag "cache"?

Reply:
No, not yet atleast. Will cover it as part of https://github.com/apache/airflow/pull/16368 or in a separate PR.

On Webserver we rely on `worker_refresh_interval` as `cache buster`.

Reply:
I added this after I manually removed dag files of queued dags which got stuck. On testing again, the dag got stuck, executor detects it and fail the task which scheduler fails too, so I will remove it.